README.md
# q2-dbbact
A [Qiime2](https://qiime2.org/) plugin for [dbBact](http://dbbact.org)

![wordcloud](https://github.com/amnona/q2-dbbact/blob/main/pics/cfs-wordcloud.jpg)
![enriched barplot](https://github.com/amnona/q2-dbbact/blob/main/pics/enriched_terms.jpg)
![heatmap](https://github.com/amnona/q2-dbbact/blob/main/pics/heatmap.jpg)
 
# Features:
* Differential abundance testing using [Calour](https://github.com/biocore/calour) rank-mean differential abundance test (with [dsFDR](https://escholarship.org/content/qt3j68q5n7/qt3j68q5n7_noSplash_e7ad1cf405f67b9cef0e5a99c1804fd5.pdf) correction).
* dbBact term enrichment from differntial abundance results of qiime2 (i.e. songbird/q2-aldex2/ancom/dacomp or the built in rank-mean test).
* Create a wordcloud of dbBact terms for a given feature table.
* Generate an interactive heatmap visualization for a feature table. The heatmap provides links to dbBact annotations for each ASV.
* Generate Venn diagram for a differential abundance result and a given dbBact term.
* Background dbBact term enrichment analysis for experiments without controls (i.e. what terms are enriched in the bacteria in a given feature table compared to all dbBact experiments of a given type).

# Examples:
## Run the q2-dbBact enrichment pipeline for a given feature table:
Our input is a feature table and a metadata file with a given column dividing our samples into two groups.

q2-dbBact will detect ASVs different between the two groups, and identify dbBact terms enriched in one of the two groups compared to the other

``` qiime dbbact enrich-pipeline --i-table cfs-merged.qza --m-metadata-file map.cfs.txt --p-field Subject --output-dir cfs-pipeline```

## Draw an interactive heatmap
This creates a zoomable heatmap with a list of dbBact annotation for each bacteria that is clicked. Useful for exploring your sequencing results and getting a feeling for what is going on (contaminations, bacterial sources, groups of samples, etc.)

Our input is a feature table and a metadata file with a given column dividing our samples into two groups.

```qiime dbbact heatmap --i-table cfs-table.qza --i-repseqs cfs-rep-seqs.qza --i-taxonomy cfs-taxonomy.qza --m-metadata-file map.cfs.txt --p-sort-field Subject --o-visualization heatmap-cfs```

![heatmap](https://github.com/amnona/q2-dbbact/blob/main/pics/heatmap.jpg)

## Draw a dbBact terms wordcloud for the set of bacteria in a feature-table
The wordcloud is created for all the bacteria in the feature table.

The output wordcloud words are dbBact terms associated with the bacteria. The word size corresponds to the F-score (recall and precision) of the term. Blue terms are positively associated (i.e. appear in COMMON/DOMINANT/HIGHER IN annotations) where as red terms (preceeded by a "-") are negatively associated (i.e. appear in LOWER IN annotations).

```qiime dbbact draw-wordcloud-vis --i-data cfs-table.qza --i-repseqs cfs-rep-seqs.qza --o-visualization wordcloud-cfs```

![wordcloud](https://github.com/amnona/q2-dbbact/blob/main/pics/cfs-wordcloud.jpg)

## Identify differentially abundant bacteria between two sample groups
q2-dbBact utilizes the non-parametric (permutation based) Calour diff_abundance() function. By default it uses a rank-mean test with dsFDR multiple hypothesis correction.

The test can also be performed as a paired test using an additional metadata pair-field (permutations are performed only between samples sharing the same pair-field value).

```qiime dbbact diff-abundance --i-table cfs-merged.qza --m-metadata-file map.cfs.txt --p-field Subject --p-alpha 0.1 --p-val1 Patient --p-val2 Control --o-diff diff-cfs-dsfdr```

## Identify and plot enriched dbBact terms between two groups of bacteria
Performed on the output of a differential-abundance test. q2-dbBact supports the following formats:
* [songbird](https://github.com/biocore/songbird)
* [ancom](https://github.com/qiime2/q2-composition)
* [q2-aldex2](https://library.qiime2.org/plugins/q2-aldex2/24/)
* dbBact diff-abundance
* any tsv file

This command identifies dbBact terms the are significantly more associated with bacteria from one group compared to the other

```qiime dbbact enrichment --i-diff diff-cfs-dsfdr.qza --p-source dsfdr --o-enriched enriched-cfs-dsfdr```

The output can be visualized (and the complete table saved) using the visualization command:

```qiime dbbact plot-enrichment --i-enriched enriched-cfs-dsfdr.qza --o-visualization barplot-enriched-cfs-dsfdr --p-labels CFS Control```

![enriched barplot](https://github.com/amnona/q2-dbbact/blob/main/pics/enriched_terms.jpg)

## Venn diagram for examining term distribution in the two groups
Input is the results of a differential abundance analysis (which provides two ASV groups - positive and negative effect size), and a dbBact term.

The venn diagram shows how many of the ASVs in each group have the term, as well as how many total dbBact ASVs have the term associated.

```qiime dbbact venn --i-diff diff-cfs-dsfdr.qza --p-terms "small village" --p-source dsfdr --p-label1 Control --p-label2 CFS --o-visualization venn-cfs-human-village```

![venn](https://github.com/amnona/q2-dbbact/blob/main/pics/venn-cfs-village.png)


CHANGELOG.md
## v1.5.0 (2023-04-19)

### Feat

- Add continuous-enrichment method for testing dbBact term enrichment without performing a prior differential abundance test (based on ASV frequencies)

## v1.4.0 (2022-05-22)

### Feat

- Support primer trimming after incomplete external primer trimming software

## v1.3.0 (2021-07-01)

### Feat

- Add better error msg for reads not matching any primer

## v1.2.0 (2021-07-01)

### Fix

- Fix bug in interpreting repseqs in _seqs_from_repseqs()

### Feat

- add term-type option for enrichment allowing annotation enrichment analysis

## v1.1.1 (2021-06-13)

### Fix

- change input parameter name for wordcloud and fix unit testing

## v1.1.0 (2021-06-12)

### Feat

- add trim-primers and primer testing in all methods/visualizations

## v1.0.3 (2021-06-01)

### Fix

- add pypi markdown support for readme

## v1.0.2 (2021-06-01)

### Fix

- fix the readme to work on pypi

## v1.0.1 (2021-06-01)

### Fix

- correct the package init file in setup.py

## v1.0.0 (2021-06-01)

### Feat

- add save pdf to venn diagram

## v0.5.0 (2021-05-30)

### Feat

- enrich-pipeline now works
- unit testing and working on pipeline

## v0.4.0 (2021-05-29)

### Feat

- add lots of new methods and visualizations

## v0.3.0 (2020-12-16)

### Feat

- Add import for new diff. abundance formats

## v0.2.0 (2020-12-06)

### Feat

- improve heatmap plugin

## v0.1.0 (2020-12-04)

### Feat

- add cluster and option to heatmap command


q2_dbbact/tests/test_dbbact.py
import unittest
import logging
from os.path import join, dirname, abspath

import pandas as pd
from pandas.testing import assert_frame_equal
import qiime2
import biom
import numpy as np

from q2_dbbact.methods import diff_abundance, enrichment


class Test_DBBact(unittest.TestCase):
	def setUp(self):
		self.test_data_dir = join(dirname(abspath(__file__)), 'data')

	def test_diff_abund(self):
		# load the data
		table = qiime2.Artifact.load(join(self.test_data_dir, 'cfs-merged.qza')).view(biom.Table)
		metadata = qiime2.Metadata.load(join(self.test_data_dir, 'map.cfs.txt'))
		table_no_repseqs = qiime2.Artifact.load(join(self.test_data_dir, 'cfs-table.qza')).view(biom.Table)
		rep_seqs = join(self.test_data_dir, 'cfs-repseqs.fasta')

		res = diff_abundance(table=table, metadata=metadata, field='Subject', random_seed=2021)
		# did we get the correct number of results?
		self.assertEqual(len(res), 53)
		# and correct number of positive?
		self.assertEqual(np.sum(res['dir'] == 0), 16)

		# integrate representative seqs
		res2 = diff_abundance(table=table_no_repseqs, metadata=metadata, field='Subject', random_seed=2021, repseqs=rep_seqs)
		# did we get the correct number of results?
		self.assertEqual(len(res), 53)
		# and correct number of positive?
		self.assertEqual(np.sum(res['dir'] == 0), 16)

		# and make sure results are the same
		assert_frame_equal(res, res2)

	def test_enrichment(self):
		repseqs = join(self.test_data_dir, 'cfs-repseqs.fasta')
		# to generate the songbird results we use:
		# QIIME 2 Plugin 'songbird' version 1.0.4 (from package 'songbird' version 1.0.4)
		# qiime songbird multinomial --i-table cfs-merged.qza --m-metadata-file map.cfs.txt --p-formula Subject --o-differentials diff-cfs-songbird.qza --o-regression-stats sb-stats.qza --o-regression-biplot sb-plot
		sb_table = qiime2.Artifact.load(join(self.test_data_dir, 'diff-cfs-songbird.qza')).view(pd.DataFrame)
		res = enrichment(diff=sb_table, source='songbird', maxid=6279, random_seed=2021)
		self.assertEqual(len(res), 439)
		self.assertEqual(np.sum(res['odif'] > 0), 97)
		self.assertEqual(res.index.values[0], 'rural community')
		self.assertEqual(res.index.values[-1], 'LOWER IN male')

		# correlation enrichment (i.e. order instead of 2 groups)
		res = enrichment(diff=sb_table, method='correlation', source='songbird', maxid=6279, random_seed=2021)
		self.assertEqual(len(res), 495)
		self.assertEqual(np.sum(res['odif'] > 0), 182)
		self.assertEqual(res.index.values[4], '-city')
		self.assertEqual(res.index.values[-1], '-thailand')

		# to generate aldex2 results we use:
		# QIIME 2 Plugin 'aldex2' version 1.14.1 (from package 'q2-aldex2' version 0+untagged.36.ga035a04.dirty)
		#  qiime aldex2 aldex2 --i-table cfs-table.qza --m-metadata-file map.cfs.txt --m-metadata-column Subject --o-differentials diff-cfs-aldex2.qza
		ad2_table = qiime2.Artifact.load(join(self.test_data_dir, 'diff-cfs-aldex2.qza')).view(pd.DataFrame)
		# we test with sig_threshold=0.5 and do correlation method since no significant results at 0.1
		res = enrichment(diff=ad2_table, repseqs=repseqs, source='aldex2', method='correlation', sig_threshold=0.5, maxid=6279, random_seed=2021)
		self.assertEqual(len(res), 365)
		self.assertEqual(np.sum(res['odif'] > 0), 101)
		self.assertEqual(res.index.values[0], 'small village')
		self.assertEqual(res.index.values[-1], '-physical activity')

		# # to generate ancom results:
		# # QIIME 2 Plugin 'composition' version 2020.6.0 (from package 'q2-composition' version 2020.6.0)
		# # qiime composition add-pseudocount --i-table cfs-merged.qza --o-composition-table cfs-pseudocount.qza
		# # qiime composition ancom --i-table cfs-pseudocount.qza --m-metadata-file map.cfs.txt --m-metadata-column Subject --output-dir diff-cfs-ancom --o-visualization diff-cfs-ancom.qzv
		# # then need to save the ancom statistical results tsv from the ancom output qzv (named ancom.tsv)
		# res = enrichment(diff_tsv=join(self.test_data_dir, 'diff-cfs-ancom-export.tsv'), source='ancom', maxid=6279, random_seed=2021)
		# print(len(res))
		# print(np.sum(res['odif'] > 0))
		# print(res.iloc[0]['odif'])
		# print(res.index.values[0])
		# self.assertEqual(len(res), 365)
		# self.assertEqual(np.sum(res['odif'] > 0), 101)
		# self.assertEqual(res.index.values[0], 'small village')
		# self.assertEqual(res.index.values[-1], '-physical activity')


q2_dbbact/utils.py
from collections import defaultdict
import re

from q2_types.feature_data import DNAFASTAFormat
import pandas as pd
import numpy as np

import calour as ca


def _iter_fasta(fp):
    '''Read fasta file into iterator.

    Fasta file must contain header line (starting with ">") and one or more sequence lines.

    Parameters
    ----------
    fp: str
        name of the fasta file

    Yields
    ------
    header: str
        the header line (without ">")
    sequence: str
        the sequence ('ACGT'). Both header and sequence are whitespace stripped.
    '''
    # skip non-header lines at beginning of file
    with open(fp, 'r') as fl:
        for cline in fl:
            if cline[0] == ">":
                title = cline[1:].rstrip()
                break
            print('Fasta file %s has no headers' % fp)
            return

        lines = []
        for cline in fl:
            if cline[0] == ">":
                yield title, ''.join(lines)
                lines = []
                title = cline[1:].strip()
                continue
            lines.append(cline.strip())
        yield title, "".join(lines)


def _seqs_from_repseqs(data: pd.DataFrame, repseqs: DNAFASTAFormat):
    '''add sequences from repseqs to the data

    Parameters
    ----------

    Returns
    -------
    data: pandas.DataFrame
        with the updated index containing sequences instead of hashes
    '''
    # check if feature ids are hashes by looking at first feature
    if len(data.iloc[0].name) == 32:
        print('replacing hashes with sequences')
        rename_ids = {}
        for chead, cseq in _iter_fasta(str(repseqs)):
            if chead in data.index:
                rename_ids[chead] = cseq
        if len(rename_ids) < len(data):
            print('Missing rep. sequences for some hashes. Found %d out of %d' % (len(rename_ids), len(repseqs)))
        data.rename(index=rename_ids, inplace=True)
    else:
        print('table does not seem to contain qiime2 hashes for ASV ids.\nFirst ASV id is %s\n' % data.iloc[0].name)
    return data


def _load_diff_abundance(diff, diff_tsv, source, ancom_stat, repseqs=None, sig_threshold=0.1):
    '''Load a diff_abundance qiime2 artifact anc convert it to a calour experiment

    Supports multiple formats (songbird / aldex2 / ancom / tsv)

    Parameters
    ----------
    diff: biom.Table or None
        the differential abundance biom table (or None)
    diff_tsv: str or None
        A tsv file name containing the diff. abundance table
    source: str
        the source for the diff. abundance output. options are:
            "songbird"
            "aldex2"
            "ancom"
            "tsv": use an external tsv file
    ancom_stat: str, optional
        name of the ancom_stat output file (for source=ancom)
    repseqs: q2_types.feature_data.DNAFASTAFormat or None
        if not None, incorporate representative sequences from this file (instead of the hashes in the table)
    sig_threshold: float
        the significance threshold for including sequences from the diff. abundance test

    Returns
    -------
    ca.AmpliconExperiment
        containing the diff. abundance features and the direction each feature is associated with
    '''
    # handle either diff (qza diffabundance artifact) or diff_tsv(tsv file resulting from diff abundance export i.e. ancom etc.)
    if diff is None and diff_tsv is None:
        raise ValueError('Need to supply either --i-diff or --p-diff-tsv differential results input file (none were supplied)')
    if diff is not None and diff_tsv is not None:
        raise ValueError('Need to supply either --i-diff or --p-diff-tsv differential results input file (both were supplied)')

    # get the differential abundance dataframe. index is sequence
    if diff is not None:
        data = diff
    else:
        try:
            data = pd.read_csv(diff_tsv, sep='\t')
            data.set_index(data.columns[0], drop=False, inplace=True)
        except Exception as e:
            print('failed to read diff_tsv file %s. Is is a valid tsv differential abundance file?' % diff_tsv)
            raise e

    # if needed, replace q2 hashes with sequences from the rep-seqs file
    if repseqs is not None:
        data = _seqs_from_repseqs(data=data, repseqs=repseqs)

    ndata = data.copy()

    # check if feature ids are hashes by looking at first feature
    if len(data.iloc[0].name) == 32:
        print('ASV IDs seem to be hashes. First id=%s' % data.iloc[0].name)
        raise ValueError('Input file contains sequence hashes instead of actual sequences.\nPlease use qiime dbbact enrichment-hash and supply the rep_seqs.qza.')

    # preprocess the dataframe to capture the correct per-feature differential abundance fields
    # these will be translated into "effect" (effect size),"pval" (p-value for differential abundance), "dir" (higher in which group) and "reject" (reject the null hypothesis of similar distribution in both groups)
    ndata['_feature_id'] = ndata.index.values
    if source == 'dsfdr':
        ndata = pd.DataFrame(data={'dir': data['Statistic'] > 0,
                                   'pval': data['raw pvalue'],
                                   'effect': data['Statistic'],
                                   'reject': data['Reject']}, index=ndata.index)
    elif source == 'songbird':
        vals = list(ndata.columns.values)
        if 'Intercept' in vals:
            vals.remove('Intercept')
        else:
            print('"Intercept" not in input data. Colums are: %s. Is it a songbird differentials file?' % vals)
            raise ValueError('File does not seem to be a songbird differentials file (no "Intercept" column)')
        vals.remove('_feature_id')
        if len(vals) > 1:
            print('More than 1 column detected: %s' % vals)
            raise ValueError('More than 1 column in songbird results (%s). Current version works only on 1 column' % vals)

        # we don't have p-values so do not reject any
        ndata = pd.DataFrame(data={'effect': ndata[vals[0]], 'reject': 1}, index=ndata.index)
    elif source == 'aldex2':
        ndata = pd.DataFrame(data={'dir': data['effect'] > 0,
                                   'pval': data['we.eBH'],
                                   'effect': data['effect'],
                                   'reject': data['we.eBH'] < sig_threshold}, index=ndata.index)
        ndata = ndata[ndata['reject'] == 1]
    elif source == 'ancom':
        # if user supplied the ancom stats file, use it to get significant (null hypothesis rejects)
        # otherwise, use the sig_threshold for the W stat
        ndata = pd.DataFrame(data={'dir': data['clr'] > 0,
                                   'effect': data['clr'],
                                   'pval': 0,
                                   'reject': data['W'] > sig_threshold}, index=ndata['id'])
        if ancom_stat is not None:
            try:
                rejects = pd.read_csv(ancom_stat, sep='\t', index_col=0)
                ndata = ndata.join(rejects)
                ndata['reject'] = ndata['Reject null hypothesis'] * 1
            except Exception as e:
                print('failed to read ancom-stat file %s. Is is a valid tsv differential abundance file?' % ancom_stat)
                raise e
    elif source == 'tsv':
        if 'pval' not in data.columns:
            raise ValueError('input tsv file does not contain "pval" column')
        if 'effect' not in data.columns:
            raise ValueError('input tsv file does not contain "effect" column')
        if 'reject' not in data.columns:
            data['reject'] = 1
        ndata = pd.DataFrame(data={'dir': data['effect'] > 0,
                                   'effect': data['effect'],
                                   'pval': data['pval'],
                                   'reject': data['reject']}, index=ndata['id'])
    else:
        raise ValueError('Unsupported source %s' % source)

    # create a new calour experiment using sequences as features and 1 sample (we don't use the sample)
    exp = ca.AmpliconExperiment(data=np.ones([1, len(ndata)]),
                                sample_metadata=pd.DataFrame({'_sample_id': ['s1']}),
                                feature_metadata=ndata,
                                sparse=False)
    return exp


def _test_exact_region(seqs, kmers={'v4': ['TACG'], 'v3': ['TGGG', 'TGAG'], 'v1': ['GACG', 'GATG', 'ATTG']}, min_fraction=0.5, ltrim=0):
    '''Test if a fasta file starts with known region k-mers

    Parameters
    ----------
    seqs: list of str
        the sequences to test
    kmers: dict of {region_name(str): [kmers(str)]}
        the kmers expected to begin each primer region
        get the values using ~/scripts/count_kmer.py -d v1
    num_reads: int
        the maximal number of reads per fileto test
    min_fraction: float
        the minimal expected fraction of reads in the region starting with the kmers (summed over all kmers of the region)
    ltrim: int, optional
        position of first nucleotide to start with. 0 to start from beginning

    Returns
    -------
    region: str or None
        the matching region or None if no region
    '''
    kmer_len = len(kmers['v4'][0])
    print('testing kmer head for region on %d sequences:' % len(seqs))
    kmer_dist = defaultdict(float)
    for cseq in seqs:
        cseq = cseq[ltrim:ltrim + kmer_len]
        for cregion, ckmers in kmers.items():
            if cseq in ckmers:
                kmer_dist[cregion] += 1

    if len(kmer_dist) > 0:
        maxregion = max(kmer_dist, key=kmer_dist.get)
        if kmer_dist[maxregion] / len(seqs) >= min_fraction:
            return maxregion
        else:
            print('No region exact match detected. Maximal region was %s but fraction of matching sequences is %f.' % (maxregion, kmer_dist[maxregion] / len(seqs)))
            return None
    print('No region exact match detected. No sequence matched any of the regions.')
    return None


def _test_embedded_primers(seqs, primers={'AGAGTTTGATC[AC]TGG[CT]TCAG': 'v1', 'CCTACGGG[ACGT][CGT]GC[AT][CG]CAG': 'v3', 'GTGCCAGC[AC]GCCGCGGTAA': 'v4'}, max_start=25, min_primer_len=10, min_fraction=0.25):
    '''Check if the sequences contain one of a given set of primers in the beginning.

    Parameters
    ----------
    seqs: list of str
        the sequences to test
    primers: dict of {primer(str): region_name(str)}, optional
        the primers to test for
    max_start: int, optional
        maximal start position for the primer (i.e. do not return if primer starts after position max_start)
    min_primer_len: int, optional
        trim primers to keep only min_primer_len last chars
    num_reads: int, optional
        the number of reads in the fasta file to process
    min_fraction: float, optional
        need at least min_fraction sequence matches in order to return the primer

    Returns
    -------
    primer: str or None
        the primer identified as maximal (if >min_fraction matches in tested reads)
    primer_name: str or None
        the name of the primer region identified
    '''
    # attach the base_dir if needed
    print('Testing %d sequences for embedded primers (%d)' % (len(seqs), len(primers)))

    # trim the primers if needed
    if min_primer_len is not None:
        print('Trimming primers before test to length %d' % min_primer_len)
        new_primers = {}
        for k, v in primers.items():
            pos = len(k)
            numchars = 0
            newp = ''
            while True:
                if numchars >= min_primer_len:
                    break
                pos = pos - 1
                if pos < 0:
                    break
                if k[pos] != ']':
                    newp = k[pos] + newp
                    numchars += 1
                    continue
                while k[pos] != '[':
                    newp = k[pos] + newp
                    pos = pos - 1
                newp = k[pos] + newp
                numchars += 1
            new_primers[newp] = v
        primers = new_primers
        print('Trimmed primers are: %s' % primers)

    # scan the files
    matches = defaultdict(float)
    for cseq in seqs:
        cseq = cseq.upper()
        for cprimer in primers.keys():
            ccseq = cseq[:max_start + len(cprimer)]
            match = re.search(cprimer, ccseq)
            if match is not None:
                matches[cprimer] += 1

    if len(matches) == 0:
        print('No embedded primers found in sequences')
        return None, None

    max_primer = max(matches, key=matches.get)
    match_fraction = matches[max_primer] / len(seqs)
    print('Found maximal matching primer: %s with enough matches (%f)' % (max_primer, match_fraction))
    if match_fraction < min_fraction:
        print('Not using primer. Maximal matches are not enough.')
        return None, None
    return max_primer, primers[max_primer]


def _trim_known_primer(seqs, fprimer, rprimer=None, length=150, remove_ambig=True, keep_primers=False):
    '''Trim until a known primer in the sequences

    Parameters
    ----------
    seqs: list of Str
        the sequences to trim ('ACGT')
    fprimer: str
        the forward primer to use ('ACGT')
    rprimer: str or None, optional
        the reverse primer to use ('ACGT') or None to not match reverse (trim only according to fprimer and length)
    length: int, optional
        the number of bases after the end of the primer to keep
    remove_ambig: bool, optional
    keep_primers: bool, optional
        If True, keep the primers trim until the beginning of the fprimer). If False, start sequences at the end of fprimer.

    Returns
    -------
    newseqs: dict, where key is the seqs_df index, and value is the trimmed sequence
    NOTE: sequences not containing the primer are not included in the output
    '''
    print('trimming %d sequences with primer %s' % (len(seqs), fprimer))
    newseqs = {}
    for ccseq in seqs:
        cseq = ccseq.upper()
        reverse_primer_seq = ''
        # find the start of the primer and output all following sequence
        try:
            match = re.search(fprimer, cseq)
            if keep_primers:
                forward_primer_seq = cseq[match.start():]
                fplen = match.end() - match.start()
            else:
                forward_primer_seq = cseq[match.end():]
        except AttributeError:
            forward_primer_seq = ''

        # if sequence can be amplified, search for reverse primer
        if forward_primer_seq != '':
            if rprimer is not None:
                # find the reverse primer match and use only up to it
                try:
                    match = re.search(rprimer, forward_primer_seq)
                    if keep_primers:
                        reverse_primer_seq = forward_primer_seq[:match.end()]
                    else:
                        reverse_primer_seq = forward_primer_seq[:match.start()]
                except AttributeError:
                    reverse_primer_seq = ''
            else:
                # reverse_primer_seq = forward_primer_seq[:length + fplen]
                reverse_primer_seq = forward_primer_seq[:]

            if reverse_primer_seq != '':
                # trim length if needed
                if length > 0:
                    if keep_primers:
                        reverse_primer_seq = reverse_primer_seq[:length + fplen]
                    else:
                        reverse_primer_seq = reverse_primer_seq[:length]
                printit = True

                if remove_ambig:
                    if 'N' in reverse_primer_seq:
                        printit = False

                if printit:
                    newseqs[ccseq] = reverse_primer_seq
    print('trimmed %d sequences, %d not found' % (len(newseqs), len(seqs) - len(newseqs)))
    return newseqs


q2_dbbact/visualizations.py
from q2_types.feature_data import DNAFASTAFormat
import biom
import pandas as pd
import numpy as np
import os
import calour as ca
import dbbact_calour
import dbbact_calour.dbbact
from typing import List

from .utils import _seqs_from_repseqs, _load_diff_abundance, _test_exact_region


def draw_wordcloud_vis(output_dir: str, table: biom.Table, repseqs: DNAFASTAFormat = None, prev_thresh: float = 0.3, focus_terms: List[str] = None):
    '''draw the wordcloud for features in the biom table and save outputs'''
    db = dbbact_calour.dbbact.DBBact()
    db.set_log_level('INFO')

    df = table.to_dataframe()

    if repseqs is not None:
        df = _seqs_from_repseqs(data=df, repseqs=repseqs)

    if len(df.iloc[0].name) == 32:
        raise ValueError('input table seems to contain hashes and not sequences. Please supply the rep-seqs.qza file.')

    exp = ca.AmpliconExperiment.from_pandas(df.T)

    # check if the experiment contains trimmed sequences. otherwise let the user know
    seqs = exp.feature_metadata.sample(n=np.min([500, len(exp.feature_metadata)])).index.values
    region = _test_exact_region(seqs)
    if region is None:
        raise ValueError('Table seems to contain untrimmed sequences. Please run qiime dbbact trim-primers on the table prior to running enrichment.')
    else:
        print('Identified region %s' % region)

    print('%d ASVs in table before prevalence filtering' % len(exp.feature_metadata))
    exp = exp.filter_prevalence(prev_thresh)
    print('%d ASVs in table remain after prevalence filtering (%f)' % (len(exp.feature_metadata), prev_thresh))
    if focus_terms is not None:
        print('Using %d focus terms: %s' % (len(focus_terms), focus_terms))
    print('getting term stats')
    fscores, recall, precision, term_count, reduced_f = db.get_wordcloud_stats(exp=exp, focus_terms=focus_terms)
    df = pd.DataFrame(data={'fscore': fscores, 'recall': recall, 'precision': precision, 'term_count': term_count, 'reduced_f': reduced_f})
    df.to_csv(os.path.join(output_dir, 'scores.tsv'), sep='\t')
    print('drawing wordcloud')
    f = db.draw_wordcloud(exp, focus_terms=focus_terms)
    f.savefig(os.path.join(output_dir, 'wordcloud.pdf'))
    f.savefig(os.path.join(output_dir, 'wordcloud.svg'))
    with open(os.path.join(output_dir, 'index.html'), 'w') as fl:
        fl.write('<html><body>\n')
        fl.write('<h1>dbBact wordcloud</h1>\n')
        fl.write('<h2>prevalence threshold=%f</h2>\n' % prev_thresh)
        if focus_terms is not None:
            fl.write('<h2>focus-terms: %s</h2>' % focus_terms)
        fl.write('<br><img src="wordcloud.svg" alt="wordcloud">\n')
        fl.write('<a href="wordcloud.pdf">Download as PDF</a><br>\n')
        fl.write('<a href="scores.tsv">Download scores as tsv</a><br>\n')
        fl.write('</body></html>')


def plot_enrichment(output_dir: str, enriched: pd.DataFrame, max_show: int = 10, max_len: int = 40, colors: List[str] = ['green', 'red'],
                    labels: List[str] = ['group1', 'group2'], enriched_exp_color: str = 'white'):
    '''Plot the enriched terms bar plot
    '''
    enriched['term'] = enriched.index.values
    numbers_kwargs = {'color': enriched_exp_color, 'weight': 'bold'}
    ax = ca.plotting.plot_enrichment(None, enriched, max_show=max_show, max_len=max_len, colors=colors, labels=labels, numbers_kwargs=numbers_kwargs)
    ax.figure.tight_layout()
    ax.figure.savefig(os.path.join(output_dir, 'enriched_terms.svg'), bbox_inches='tight')
    ax.figure.savefig(os.path.join(output_dir, 'enriched_terms.pdf'), bbox_inches='tight')
    enriched.to_csv(os.path.join(output_dir, 'enriched_terms.tsv'), sep='\t')
    with open(os.path.join(output_dir, 'index.html'), 'w') as fl:
        fl.write('<html><body>\n')
        fl.write('<h1>dbBact enriched terms</h1>\n')
        fl.write('<img src="enriched_terms.svg" alt="enriched terms barplot"><br><br>\n')
        fl.write('<a href="enriched_terms.pdf">Download as PDF</a><br>\n')
        fl.write('<a href="enriched_terms.tsv">Download enriched terms as tsv</a><br>\n')
        fl.write('</body></html>')


def heatmap(output_dir: str, table: biom.Table, metadata: pd.DataFrame, sort_field: str = None, cluster: bool = True, min_abundance: float = 10, normalize: bool = True,
            taxonomy: pd.DataFrame = None, repseqs: DNAFASTAFormat = None):
    ca.set_log_level('INFO')
    print('creating experiment')
    metadata = metadata.to_dataframe()
    samples = table.ids(axis='sample')
    features = table.ids(axis='observation')
    # combine the taxonomy
    feature_metadata = pd.DataFrame(index=features, columns={'_feature_id': features})
    if taxonomy is not None:
        print('adding taxonomy')
        feature_metadata['taxonomy'] = taxonomy['Taxon']
    else:
        feature_metadata['taxonomy'] = 'NA'
    # use repseqs if supplied
    if repseqs is not None:
        print('converting hashes to sequences using repseqs file')
        feature_metadata = _seqs_from_repseqs(feature_metadata, repseqs)

    metadata = metadata.filter(items=samples, axis='index')
    exp = ca.AmpliconExperiment(data=table.transpose().matrix_data, sample_metadata=metadata, feature_metadata=feature_metadata, sparse=False)
    print('created experiment %r' % exp)
    if normalize:
        print('normalizing to 10000 reads/sample')
        exp = exp.normalize(10000)
    if min_abundance > 0:
        print('removing sequences with sum abundance < %f' % min_abundance)
        exp = exp.filter_sum_abundance(min_abundance)
        print('remaining experiment %r' % exp)
    if sort_field is not None:
        print('sorting by field %s' % sort_field)
        exp = exp.sort_samples(sort_field)
    if cluster:
        print('clustering sequences')
        exp = exp.cluster_features(10)
    print('generating heatmap to output file %s' % output_dir)
    exp.export_html(output_file=os.path.join(output_dir, 'index.html'), sample_field=sort_field, feature_field='taxonomy')


def venn(output_dir: str, terms: List[str], diff: pd.DataFrame = None, repseqs: DNAFASTAFormat = None, diff_tsv: str = None, source: str = 'dsfdr', sig_threshold: float = 0.1, ancom_stat: str = None,
         label1: str = 'Negative', label2: str = 'Positive', set_colors: List[str] = ['red', 'green', 'mediumblue'], max_size: int = 1000):
    ca.set_log_level('INFO')
    db = dbbact_calour.dbbact.DBBact()
    db.set_log_level('INFO')

    # load the differetial abundance table into a calour experiment
    exp = _load_diff_abundance(diff=diff, diff_tsv=diff_tsv, source=source, ancom_stat=ancom_stat, repseqs=repseqs, sig_threshold=sig_threshold)
    # exp.feature_metadata['_calour_direction'] = exp.feature_metadata['dir']
    exp.feature_metadata['_calour_direction'] = exp.feature_metadata['dir'].replace({0: label1, 1: label2}, inplace=False)

    # check if the experiment contains trimmed sequences. otherwise let the user know
    seqs = exp.feature_metadata.sample(n=np.min([500, len(exp.feature_metadata)])).index.values
    region = _test_exact_region(seqs)
    if region is None:
        raise ValueError('Table seems to contain untrimmed sequences. Please run qiime dbbact trim-primers on the table prior to running enrichment.')
    else:
        print('Identified region %s' % region)

    f = db.plot_term_venn_all(terms, exp, max_size=max_size, set_colors=set_colors)
    f.savefig(os.path.join(output_dir, 'venn.svg'))
    f.savefig(os.path.join(output_dir, 'venn.pdf'))
    with open(os.path.join(output_dir, 'index.html'), 'w') as fl:
        fl.write('<html><body>\n')
        fl.write('<h1>dbBact venn for terms %s</h1>\n' % ','.join(terms))
        fl.write('<img src="venn.svg" alt="term venn diagram"><br><br>\n')
        fl.write('<a href="venn.pdf">Download as PDF</a><br>\n')
        fl.write('</body></html>')
